pkg/importer/data_processor.go

```diff
@@ -0,0 +1,211 @@
+	// Resize the disk image, this is only needed when the target contains a file system (block device do not need a resize)
+	Resize ProcessingPhase = "Resize"
+	// Complete is the phase where the entire process completed successfully and we can exit gracefully.
+	Complete ProcessingPhase = "Complete"
+	// Error is the phase in which we encountered an error and need to exit ungracefully.
+	Error ProcessingPhase = "Error"
+)
+
+// ErrRequiresScratchSpace indicates that we require scratch space.
+var ErrRequiresScratchSpace = fmt.Errorf("Scratch space required and none found")
+
+// DataProviderInterface is the interface all data providers should implement.
+type DataProviderInterface interface {
+	// Info() is called to get initial information about the data.
+	Info() (ProcessingPhase, error)
+	// Transfer() is called to transfer the data from the source to a temporary location.
+	Transfer(path string) (ProcessingPhase, error)
+	// Process() is called to do any special processing before giving the url to the data back to the processor
+	Process() (ProcessingPhase, error)
+	// Geturl returns the url that the data processor can use when converting the data.
+	GetURL() *url.URL
+	// Close closes any readers or other open resources.
+	Close() error
+}
+
+// DataProcessor holds the fields needed to process data from a data provider.
+type DataProcessor struct {
+	// currentPhase is the phase the processing is in currently.
+	currentPhase ProcessingPhase
+	// provider provides the data for processing.
+	provider DataProviderInterface
+	// destination file. will be DataDir/disk.img if file system, or a block device (if a block device, then DataDir will not exist).
+	dest string
+	// dataDir path to target directory if it contains a file system.
+	dataDir string
+	// scratchDataDir path to the scratch space.
+	scratchDataDir string
+	// requestImageSize is the size we want the resulting image to be.
+	requestImageSize string
+	// available space is the available space before downloading the image
+	availableSpace int64
+}
+
+// NewDataProcessor create a new instance of a data processor using the passed in data provider.
+func NewDataProcessor(dataProvider DataProviderInterface, dest, dataDir, scratchDataDir, requestImageSize string) *DataProcessor {
+	dp := &DataProcessor{
+		currentPhase:     Info,
+		provider:         dataProvider,
+		dest:             dest,
+		dataDir:          dataDir,
+		scratchDataDir:   scratchDataDir,
+		requestImageSize: requestImageSize,
+	}
+	// Calculate available space before doing anything.
+	dp.availableSpace = dp.calculateTargetSize()
+	return dp
+}
+
+// ProcessData is the main processing loop.
+func (dp *DataProcessor) ProcessData() error {
+	var err error
+	defer dp.provider.Close()
+	if util.GetAvailableSpace(dp.scratchDataDir) > int64(0) {
+		defer CleanDir(dp.scratchDataDir)
+	}
+	for dp.currentPhase != Complete {
+		switch dp.currentPhase {
+		case Info:
+			dp.currentPhase, err = dp.provider.Info()
+		case Transfer:
```

Comment:
Suggest to unite Transfer and TransferTarget into Transfer and pass both dataDir and SratchDataDir to each data_source file, so each specific dataSource will decide how and when to use dataDir or ScratchDataDir. 
In additiona, each data source can cleanup after itself on provider.Close(), so no need to call CleanDir() when it is not used

Reply:
I disagree, there is no need for 2 parameters since we are either transferring to the scratch space, or to the target space but not both. So if I pass 2 parameters one will always be empty. I also don't really want the data source to worry too much about where to write the data. The data source job is to collect the data, and then write it to where the process wants it to write.

Reply:
1. in all data sources scratch space is either required or not, but all of them eventually will copy the data to /data/disk.img
2. In fact in all those data sources where scratch is required there are validations that scratch space is provided, and if not provided then a dedicated error ErrRequiresScratchSpace is returned. So in fact data source is aware of which space it needs and when it uses it. It might be more consistent to validate scratch space in one place (either data_processor or specific data_source. I personally vote for data_source) By the way, scratch space path can be provided on data_source creation and not on Transfer, in such case it is passed only to those data_sources that can potentially require it.

Reply:
1. Not true, in particular archive content type will write to the /data directory one or more files of whatever name is in the archive. Also in the happy paths scratch space is not required, in fact the data source doesn't even write any data anywhere, it all happens in the conversion phase.
2. You are right the validation for scratch space is knowledge the data source shouldn't have, it should check if the path provided is valid or not, and return an error based on that, then the data processor should determine based on what it provided the transfer function, if it should return the scratch space error. I will fix that, I took a short cur when I shouldn't have.

As for setting the data/scratch space when creating the data source, yes I can do that, but why would I want to. I want the data source to concentrate on its purpose, getting the data. The data processor decides where that data goes.

Reply:
Right that is sort of the intention, even though we don't have any of that right now. But a good example of something we could potentially do is. Lets say we always give each pod emptyDir volume (host space). And we determine we need scratch space, if the info phase decided that the size of the incoming data is small enough, the data processor might decided to go directly to transfer with the emptyDir instead of killing the pod with the scratch space exit code.

This is just some stuff I am thinking of of the top of my head, but I think there is value in having the items separated between the processor and the data source.